### DISCUSSION TITLE
Custom WebGL particle/flow layer

### DISCUSSION THREAD

**Discussion by kingkong-cmd** [openlayers/openlayers#16196] [Show and tell]
So, I have been building a very crude version of a custom flowlayer for wind illustration, and it's starting to come together, ref img below. Hoping to get some advice on how to proceed.

So far I have a background WebGLTileLayer that creates a continuous heatmap showing wind strength, using OL styling options. This works perfectly. Then I have used a TileLayer on top for the particle animation, drawing it onto a canvas in the loader function, and return the canvas to the TileLayer. Both layers are based on data encoded in RGBA in png files.

I am now looking for advice on how to implement more properly with suitable layers, sources, renderers etc (I find it really hard to understand how all the parts work together from reading OL source):

1. What would be the best Layer to use/extend for this usecase?

2. The actual animation is now happening through a framestate.animate = true in a forever postrender event loop. Is this a correct way of telling WebGL to keep drawing? Feels like it requires some compute, but might be the WebGL process itself, and not the framestate.animate = true loop.
 
3. How should I split data preparation vs WebGL stuff between source/layer? Do i prepare data in a source, return it to a layer, and then do all the WebGL and rendering there? WebGL functions need access to the map bounds so that i can resize the canvas/layer on zoom&scroll. This is not available in the source/loader function, but is available via framestate from render function in layer.

```
wind_source = new DataTileSource({
    interpolate: true,
    maxZoom: 0,
    loader: async function (z, x, y) {

        // PSEUDOCODE:
        // Create a canvas with hard-coded witdth and height (cannot access framestate in loader function)
        // Functions for getting all the wind data from png (probably belongs in the loader function)
        // Functions for WebGL to draw on the canvas with shaders, textures, framebuffers and the whole shabang (Should be moved to layer render function?)

         function render() {
	      drawScreen()
	      updateParticlesPositions();
	      requestAnimationFrame(render);
        }
	          
        // Start the animation loop
        render();
        return canvas
}

testlayer = new TileLayer({
    source: wind_source,
})

map.addLayer(testlayer);

testlayer.on('postrender', function (event) {
	event.frameState = true
})
```

![image](https://github.com/user-attachments/assets/16d99734-1994-4af3-9353-3fc33d8869cd)

**Comment by kingkong-cmd** [comment]
Ok, I finally managed to make something. I used a custom canvaslayer and rewrote the render function to do all the webgl stuff, and then the animation runs by calling requestAnimationFrame() which again calls map.render(). I have no idea what I am doing, so if anybody is reading I would like to know if this is an anti-pattern. 

Also, still haven't figured out how to attach the layer to a Source - now i'm just reading the png directly from the render function.

A feature, different than the Flowlayer, which reprojects the image into webmercator prior to putting it into the texture, I made the shader calculate the lat/lon for the particle, so that it can look up the vectors directly from the un-reprojected png just using the lat/lon. I think this may be better. Not sure.

Also, I have separated the color into a separate layer. I thing this works better than combining color and flow, especially when moving and zooming the map. Ideally I want to clear the whole flow layer when moving the map, and redraw when it stops, but not sure how.
 
[Screencast from 18. sep. 2024 kl. 01.50 +0200.webm](https://github.com/user-attachments/assets/ad6df9a5-6edd-42ee-b031-a7d7a433f683)

**Reply by kingkong-cmd** [reply]
Managed to figure out how to change the underlying data for the flowlayer by adding some code to the subclassed layer where it fetches new pngs from urls, so now I can use an input slider thing to select underlying data. Problem now is that the flow layer renders behinds all the other layers, but over the graticule overlay, so that means transitions are ugly as hell, and no graticule. Tried playing with z-indexes but doesn't seem to work.
 
[Screencast from 19. sep. 2024 kl. 00.43 +0200.webm](https://github.com/user-attachments/assets/fa749b83-4a36-4e84-aced-c6cb83752838)

**Comment by kingkong-cmd** [comment]
Figured it out. Since I am creating my own canvas in the custom Layer class I need to set some css for it not to mess up all the layers. Now, the graticule becomes visible again, and I can change the color of the wind particles in the fragment shader since it is not hidden behind the color layer anymore. Though the slightly grayish particles looked better.  

```
this.canvas = document.createElement('canvas');
this.canvas.classList.add("ol-layer")
this.canvas.setAttribute("style", "position: absolute; left: 0px;")
```
[Screencast from 19. sep. 2024 kl. 03.59 +0200.webm](https://github.com/user-attachments/assets/9c8a1a27-4b38-4f84-af06-9b13e74a1c66)

**Comment by robitaillej** [comment]
Hi! The streams look great! I'm really interested on how you managed to do this. Do you have a wokring code sample? 
I am perticularly interested in what kind of source data is needed to do that. I am usually rendering raster data with geoserver through wms calls. So I'm not sure how I could adapt from this. I was curious on how you did it.
Thanks for any information!

**Reply by kingkong-cmd** [reply]
Hi,

It was a bit of a process since I started at zero, knowing almost nothing about any of this. Unfortunately no code to share as I'm looking to port to mapbox or something due to almost non-existent community support for OL.

Anyway, the source data is grib files with wind forecast data, which you get from the american weather service NOAA. These contain gridded data for the whole globe with wind data for each 0.25 degree grid point. Just reading these files is a whole conundrum because you need special libs. I use xarray with a new grib reader called gribberish (https://github.com/mpiannucci/gribberish), all in python.

Then you need to get this data from the backend to the front end which again is not straight forwards because you need to transfer in a efficient format (file size becomes big if not efficient). The most normal thing to do is to encode the data into png files, where each pixel of the png represents a point in the grib file.

Then, on the front-end (or in the OL application or whatever you want to call it), you need to decode the png file into javascript data type, e.g. arrays, and then transfer the data into webGL to do all the rendering. This was super complicated since I had no prior concept of it, and when coding webGL you cannot debug anything. Combine this with trying to understand how to interact with OL without being able to ask for help, and it becomes quite complex. 

So would actually not recommend unless you already understand how OL works under the hood. I spent too much time trying to read source to understand basics, which I still don't.

The OL guys did create this new layer which might be of interest https://openlayers.org/en/latest/examples/wind.html.

**Reply by rasa2k** [reply]
Hi kingkong-cmd,
Thanks for sharing. I'm impressed how far you managed to come. I'm not planning on switching away from Openlayers and I'm very keen on seeing more of your code if possible?

**Reply by robitaillej** [reply]
Hello,
Thanks kingkong-cmd for you reply. 
I got some time this week to work on this and I managed to arrive to a somewhat ok result with the experimental Flow layer from Openlayers. I still have a lot of work to do to make it better, but its still something.
The thing I wanted to do is to use one of our GeoServer layer as data source. [This codepen example](https://codepen.io/robitaillej/pen/NPKyxzP) is a somewhat derived version of the OL example.

**Reply by kingkong-cmd** [reply]
> Hello, Thanks kingkong-cmd for you reply. I got some time this week to work on this and I managed to arrive to a somewhat ok result with the experimental Flow layer from Openlayers. I still have a lot of work to do to make it better, but its still something. The thing I wanted to do is to use one of our GeoServer layer as data source. [This codepen example](https://codepen.io/robitaillej/pen/NPKyxzP) is a somewhat derived version of the OL example.

Wow, great work.

I did not quite like the Flowlayer, hence why I built my own. Still WIP, but I put it [here](https://weatherrouting.online/) while I work on it, in case you want to have a look. Not possible to read code though since minified.

Suddenly people are showing up to discuss, so maybe won't migrate to Mapbox after all. Have been looking into ways to optimize size of wind data going from server to client lately, but seems almost impossible to compress more than encoding data into pngs => meaning wind strength only, is apprx 0.5 MB for whole world with 0.25 degr resolution, and strength + direction is 1 MB for the same. Tiling the data acted weirdly on the tile edges.

**Reply by kingkong-cmd** [reply]
> Hi kingkong-cmd, Thanks for sharing. I'm impressed how far you managed to come. I'm not planning on switching away from Openlayers and I'm very keen on seeing more of your code if possible?

Hi,

Unfortunately don't have code to share right now. May share something at a later stage. @robitaillej 's example combined with my description to him can give you some pointers on how to approach, but he uses the Flowlayer, hence doesn't show the webGL stuff which is the tricky part.

**Reply by robitaillej** [reply]
> > Hello, Thanks kingkong-cmd for you reply. I got some time this week to work on this and I managed to arrive to a somewhat ok result with the experimental Flow layer from Openlayers. I still have a lot of work to do to make it better, but its still something. The thing I wanted to do is to use one of our GeoServer layer as data source. [This codepen example](https://codepen.io/robitaillej/pen/NPKyxzP) is a somewhat derived version of the OL example.
> 
> Wow, great work.
> 
> I did not quite like the Flowlayer, hence why I built my own. Still WIP, but I put it [here](https://weatherrouting.online/) while I work on it, in case you want to have a look. Not possible to read code though since minified.
> 
> Suddenly people are showing up to discuss, so maybe won't migrate to Mapbox after all. Have been looking into ways to optimize size of wind data going from server to client lately, but seems almost impossible to compress more than encoding data into pngs => meaning wind strength only, is apprx 0.5 MB for whole world with 0.25 degr resolution, and strength + direction is 1 MB for the same. Tiling the data acted weirdly on the tile edges.

The result is quite nice! The OL community seems a little bit active on the subject. I found some interesting info on this [pull request discussion](https://github.com/openlayers/openlayers/pull/14491). The Flow layer seems to be still in development.

**Reply by robitaillej** [reply]
Hello again kingkong-cmd, 
I'm starting digging this thing again, and also trying to figure out if I should move towards mapbox or not. Just so I understand [your example](https://weatherrouting.online/) correctly, is the code made with custom openlayers functions while the data is stored on mapbox? If the code is with openlayers, i'd really like to see how you managed to do something like that, and if you think you'll want to share your code at some point. Because, as you said it, the Flowlayer from openlayers is still in development and does not really fit every scenarios. And I'd really like to stick with openlayers!

**Reply by kingkong-cmd** [reply]
> Hello again kingkong-cmd, I'm starting digging this thing again, and also trying to figure out if I should move towards mapbox or not. Just so I understand [your example](https://weatherrouting.online/) correctly, is the code made with custom openlayers functions while the data is stored on mapbox? If the code is with openlayers, i'd really like to see how you managed to do something like that, and if you think you'll want to share your code at some point. Because, as you said it, the Flowlayer from openlayers is still in development and does not really fit every scenarios. And I'd really like to stick with openlayers!

First, I think it's the same complexity with both OL and mapbox. Only difference is with mapbox you will be able to ask questions underways because there are people there who will help.

I haven't stored any data on mapbox, everything is on my server. I download wind data, then encode the wind data into png for compression (using python), then those pngs are fetched by the client when the website is used, and on the client I use a regular OL WebGLTileLayer for the colors, and then a OL CanvasLayer for all the webgl drawing of the wind particles. The CanvasLayer is almost fully customised. The webGL stuff was complicated on its own, but what bugged me tho most was figuring out how to make the CanvasLayer interact with OL. This is what is difficult when there is no community.

I can probably share something someday, but don't wait for it. If you decide to try yourself, I can try to assist, but I am no expert.

**Reply by Windsurfice** [reply]
> First, I think it's the same complexity with both OL and mapbox. Only difference is with mapbox you will be able to ask questions underways because there are people there who will help.
> 
> I haven't stored any data on mapbox, everything is on my server. I download wind data, then encode the wind data into png for compression (using python), then those pngs are fetched by the client when the website is used, and on the client I use a regular OL WebGLTileLayer for the colors, and then a OL CanvasLayer for all the webgl drawing of the wind particles. The CanvasLayer is almost fully customised. The webGL stuff was complicated on its own, but what bugged me tho most was figuring out how to make the CanvasLayer interact with OL. This is what is difficult when there is no community.
> 
> I can probably share something someday, but don't wait for it. If you decide to try yourself, I can try to assist, but I am no expert.

Have you tested you flow setup on Mali GPU's?
With the Openlayer version I have this strange pattern: [https://github.com/openlayers/openlayers/issues/16432](https://github.com/openlayers/openlayers/issues/16432)

I'm building my website on the Openlayers version and this is what I got so far. [https://windsurfice.com/nl/flow-map](https://windsurfice.com/nl/flow-map)
But the Mali GPU is killing it at the moment. Here in the Netherlands most Samsung mobile phones have an Mali GPU so they can't see the website correct.

**Reply by kingkong-cmd** [reply]
> Have you tested you flow setup on Mali GPU's?

I have not, but if you go to the website I linked to with a Mali GPU device I guess you will quickly see if it works or not. I believe my particles require use of webgl2 though.

> With the Openlayer version I have this strange pattern: #16432

These issues are way above my pay grade, but I noticed my wind particles became all wonky tonky on mobile when I used `precision mediump float;` instead of `precision highp float;` in my shaders. I see there is some code in the flowlayer (ref link below) dealing with this, but the code is close to unreadable for me, so not sure what is being used in the end. Not sure how, but might be worth trying to force use of highp float.

https://github.com/search?q=repo%3Aopenlayers%2Fopenlayers+precision+highp+float&type=code

**Reply by jahow** [reply]
> I see there is some code in the flowlayer (ref link below) dealing with this, but the code is close to unreadable for me, so not sure what is being used in the end. Not sure how, but might be worth trying to force use of highp float.

Shaders in OpenLayers will use `precision highp float;` if available, and fallback to `mediump` otherwise. If you force `highp` you might get compilation errors on some mobile devices; falling back to mediump gives the opportunity to still render something, even though it might be a bit degraded.

**Reply by kingkong-cmd** [reply]
> > I see there is some code in the flowlayer (ref link below) dealing with this, but the code is close to unreadable for me, so not sure what is being used in the end. Not sure how, but might be worth trying to force use of highp float.
> 
> Shaders in OpenLayers will use `precision highp float;` if available, and fallback to `mediump` otherwise. If you force `highp` you might get compilation errors on some mobile devices; falling back to mediump gives the opportunity to still render something, even though it might be a bit degraded.

Ok, that's good to know. I just googled to find out that `#ifdef GL_FRAGMENT_PRECISION_HIGH` is actually checking this exact thing. Then it probably has nothing to do with this, unless maybe these Mali devices only support mediump, and then it just renders nonsense like it did for me when I tried forcing mediump on an Adreno device.

**Comment by Windsurfice** [comment]
Your example doesn't load the particles and color layer on my Samsung mobile. I will try to debug later. 
As for highp I checked this on my phone and it does support it but still it renders the wrong pattern.
